containerd-shim-spin/src/engine.rs

```diff
@@ -38,21 +42,102 @@ impl TriggerHooks for StdioTriggerHook {
+            AppSource::Oci => write!(f, "Oci"),
+        }
+    }
+}
+
 impl SpinEngine {
-    fn app_source(&self) -> Result<PathBuf> {
-        spin_common::paths::resolve_manifest_file_path("/spin.toml")
+    async fn app_source(&self, ctx: &impl RuntimeContext, cache: &Cache) -> Result<AppSource> {
+        match ctx.wasm_layers() {
+            [] => Ok(AppSource::File(
+                spin_common::paths::resolve_manifest_file_path("/spin.toml")?,
+            )),
+            layers => {
+                info!(
+                    " >>> configuring spin oci application {}",
+                    ctx.wasm_layers().len()
+                );
+
+                for artifact in layers {
+                    match artifact.config.media_type() {
+                        MediaType::Other(name)
+                            if name == spin_oci::client::SPIN_APPLICATION_MEDIA_TYPE =>
+                        {
+                            let path = PathBuf::from("/spin.json");
+                            log::info!("writing spin oci config to {:?}", path);
+                            File::create(&path)
+                                .context("failed to create spin.json")?
+                                .write_all(&artifact.layer)
+                                .context("failed to write spin.json")?;
+                        }
+                        MediaType::Other(name)
+                            if name == "application/vnd.wasm.content.layer.v1+wasm" =>
+                        {
+                            log::info!(
+                                "writing artifact config to cache, near {:?}",
+                                cache.manifests_dir()
+                            );
+                            cache
+                                .write_wasm(&artifact.layer, &artifact.config.digest())
+                                .await?;
+                        }
+                        _ => {}
+                    }
+                }
+                Ok(AppSource::Oci)
+            }
+        }
     }
 
-    fn resolve_app_source(&self, app_source: PathBuf) -> Result<ResolvedAppSource> {
-        Ok(ResolvedAppSource::File {
-            manifest_path: app_source.clone(),
-            manifest: spin_manifest::manifest_from_file(app_source)?,
-        })
+    async fn resolve_app_source(
+        &self,
+        app_source: AppSource,
+        cache: &Cache,
+    ) -> Result<ResolvedAppSource> {
+        let resolve_app_source = match app_source {
+            AppSource::File(source) => ResolvedAppSource::File {
+                manifest_path: source.clone(),
+                manifest: spin_manifest::manifest_from_file(source.clone())?,
+            },
+            AppSource::Oci => {
+                let working_dir = PathBuf::from("/");
+                let loader = spin_oci::OciLoader::new(working_dir);
+
+                // TODO: what is the best way to get this info? It isn't used only saved in the locked file
```

Comment:
Why is this needed?

Reply:
it is part of the API to load OCI.  https://github.com/fermyon/spin/blob/40bbead627ef6c8f61706516e5f11f1122a33f21/crates/oci/src/loader.rs#L53-L58

The vaule is stored as metadata but not used anywhere.  We could probably pass this info along to the shim somehow or maybe make it optional in spin?